Avoid tid fetch in zedstoream_tuple_satisfies_snapshot()

soumyadeep2007 opened the pull request:
This addresses the following TODO in `zedstoream_tuple_satisfies_snapshot()`:
```
/*
 * TODO: we didn't keep any visibility information about the tuple in the
 * slot, so we have to fetch it again. A custom slot type might be a
 * good idea..
 */
```
We have kept visibility information in `ZedstoreTupleTableSlot` since
1d996afef9. We now make use of that info to call
`zs_SatisfiesVisibility()` instead of having to scan. To call this, we do
need the undoPtr for the tuple. So, we add this undoPtr to the slot,
wherever we supply the slot's visibility information.

soumyadeep2007 committed:
Avoid tid fetch in zedstoream_tuple_satisfies_snapshot()

This addresses the following TODO in zedstoream_tuple_satisfies_snapshot():

	/*
	 * TODO: we didn't keep any visibility information about the tuple in the
	 * slot, so we have to fetch it again. A custom slot type might be a
	 * good idea..
	 */
We have kept visibility information in ZedstoreTupleTableSlot since
1d996afef9. We now make use of that info to call
zs_SatisfiesVisibility() instead of having to scan. To call this, we do
need the undoPtr for the tuple. So, we add this undoPtr to the slot,
wherever we supply the slot's visibility information.

soumyadeep2007 commented:
The main concern I have for this PR is the need to add the undo pointer to the `ZedstoreTupleTableSlot`. It does increase the memory requirement for the slot significantly and the case we are covering (`zedstoream_tuple_satisfies_snapshot()`) isn't a common one.

d commented:
I'd suggest you find an alternative title to this PR. "TidScan" is usually understood as something else ...

hlinnaka commented:
LGTM. I wonder if it would be more handy to store the undoptr in the ZSUndoSlotVisibility struct itself. Looks like you're passing around the two of them together a lot.

hlinnaka commented:
> It does increase the memory requirement for the slot significantly and the case we are covering (zedstoream_tuple_satisfies_snapshot()) isn't a common one.

I'm not too worried about that. A TupleTableSlot is a pretty large struct anyway, 8 more bytes (or 16?) shouldn't make much difference.

soumyadeep2007 committed:
Encapsulate UNDO pointers in visibility info.

Before this commit, we have had to pass around UNDO pointers in parallel
with visibility information.

soumyadeep2007 commented:
> LGTM. I wonder if it would be more handy to store the undoptr in the ZSUndoSlotVisibility struct itself. Looks like you're passing around the two of them together a lot.

Yup good call! I refactored that as a separate commit.
